core/sail/memory/src/test/java/org/eclipse/rdf4j/sail/memory/QueryPlanRetrievalTest.java

```diff
@@ -619,4 +618,94 @@ public void testTimeout() {
 
 	}
 
+	@Test
+	public void testAlternatePath() {
+		SailRepository sailRepository = new SailRepository(new MemoryStore());
+		try (SailRepositoryConnection connection = sailRepository.getConnection()) {
+			connection.begin();
+			for (int i = 0; i < 10; i++) {
+				connection.add(vf.createBNode(i + ""), RDF.TYPE, vf.createBNode((i + 1) + ""));
+				connection.add(vf.createBNode(i + ""), RDF.TYPE, vf.createBNode((i - 1) + ""));
+
+				connection.add(vf.createBNode(i + ""), RDF.TYPE, vf.createBNode((i + 2) + ""));
+				connection.add(vf.createBNode(i + ""), RDF.TYPE, vf.createBNode((i - 2) + ""));
+			}
+
+			connection.add(vf.createBNode("0"), RDFS.LABEL, vf.createLiteral("label"));
+
+			connection.commit();
+		}
+
+		try (SailRepositoryConnection connection = sailRepository.getConnection()) {
+			Query query = connection.prepareTupleQuery(String.join("\n", "",
+					"PREFIX rdfs:<http://www.w3.org/2000/01/rdf-schema#>",
+					"select * where {",
+					"	?a (a|^a)* ?c. ?c (a|^a) ?b .",
+					"	?b rdfs:label ?type .",
+					"}"));
+
+			query.setMaxExecutionTime(1);
+
+			String actual = query.explain(Explanation.Level.Executed).toString();
+
+			assertEquals("Projection (resultSizeActual=84)\n" +
```

Comment:
I don't particularly like the notion of using the string representation of the query plan as the thing to test on. Far too brittle in my opinion - this test would easily break if we made any changes to query processing, even if those changes have nothing to do with alternative paths.

Reply:
Fair enough. Let me just use this test to verify your fix, and we can delete both this PR and the branch after.